onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,695 @@
+                pendingElementCount,
+                1,
+                64,
+                dispatchSizeX,
+                pendingElementCount
+            );
+
+            // Set root constants
+            commandList->SetComputeRoot32BitConstants(
+                2, // root parameter index
+                21, // Constant count
+                &constants,
+                0 // offset
+            );
+
+            commandList->Dispatch(dispatchSizeX, 1, 1);
+        }
+
+        commandList->ResourceBarrier(2, uav_barriers);
+    }
+
+    void BluesteinZChirp(
+        ID3D12Resource* /*inputResource*/,
+        ID3D12Resource* /*outputResource*/,
+        ID3D12GraphicsCommandList* /*commandList*/)
+    {
+        ORT_THROW_HR(E_NOTIMPL);
+    }
+};
+
+struct DFTShapeInferrer : public WRL::Base<IMLOperatorShapeInferrer>
+{
+    STDMETHOD(InferOutputShapes)(IMLOperatorShapeInferenceContext* context) noexcept
+    {
+        try
+        {
+            int64_t axis;
+            ORT_THROW_IF_FAILED(context->GetAttribute("axis", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&axis)));
+            int64_t isInverseInt;
+            ORT_THROW_IF_FAILED(context->GetAttribute("inverse", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isInverseInt)));
+            int64_t isOnesidedInt;
+            ORT_THROW_IF_FAILED(context->GetAttribute("onesided", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isOnesidedInt)));
+            bool isOnesided = static_cast<bool>(isOnesidedInt);
+            bool inverse = static_cast<bool>(isInverseInt);
+
+            if (inverse && isOnesided) {
+                throw new std::exception("isOnesided and inverse attributes cannot be enabled at the same time");
+            }
+
+            uint32_t rank;
+            ORT_THROW_IF_FAILED(context->GetInputTensorDimensionCount(0, &rank));
+            if (rank == 0) {
+                // If no shape is available for the input, skip shape inference...
+                throw;
+            }
+
+            auto axisIdx = OperatorHelper::HandleNegativeAxis(static_cast<int32_t>(axis), rank);
+
+            // if (context->IsInputValid(1))
+            // {
+            //     // If dft_length is specified, then we should honor the shape.
+            //     // Set the output dimension to match the dft_length on the axis.
+            //     // If onesided this will be adjusted later on.
+            //     // However, this is not supported as of now, so do nothing...
+            // }
+
+            // In general the output shape will match the input shape exactly
+            // So initialize the output shape with the input shape
+            std::vector<uint32_t> inputDims(rank);
+            ORT_THROW_IF_FAILED(context->GetInputTensorShape(0, rank, inputDims.data()));
+            auto outputDims = inputDims;
+            outputDims.back() = 2;
```

Comment:
>2; [](http://example.com/codeflow?start=32&length=2)

What does the magic value 2 mean? Is it because of real and imaginary numbers? If so propose comment:

= 2; // The last dimension has complex imaginary and real components